synapse/storage/devices.py

```diff
@@ -634,19 +641,42 @@ def _mark_as_sent_devices_by_remote_txn(self, txn, destination, stream_id):
         txn.execute(sql, (destination, stream_id,))
 
     @defer.inlineCallbacks
-    def get_user_whose_devices_changed(self, from_key):
+    def get_user_whose_devices_changed(self, user_id, from_key):
         """Get set of users whose devices have changed since `from_key`.
         """
         from_key = int(from_key)
         changed = self._device_list_stream_cache.get_all_entities_changed(from_key)
         if changed is not None:
-            defer.returnValue(set(changed))
+            changed_set = set(changed)
+            if self._attestation_stream_cache.has_entity_changed(user_id, from_key):
```

Comment:
I'm a bit confused about why this bit is conditional here, but in the other branch of the function it isn't.